### ISSUE TITLE
[graph_trainer] SAR omits in-place writes to recomputed buffers, producing incorrect gradients

### ISSUE THREAD

**Issue by johnjunjun7** [pytorch/torchtitan#4688]
### Bug description

`selective_activation_remat_pass` can silently produce incorrect gradients when a recomputed buffer is modified by a schema-correct in-place operator that returns no value.

The pass follows dataflow producer edges. For a functional `autograd.Function` implemented as `clone(input) -> in-place custom op on the clone -> return clone`, the write has no dataflow output. Recomputing the clone without replaying the write gives backward the pre-mutation value.

The CPU reproduction below computes `loss = (x + 1)^2` at `x = 2`:

- Before the pass: loss = 9, gradient = 6 (correct).
- After the pass: loss = 9, gradient = 4 (incorrect).
- The transformed graph passes `graph.lint()` and executes without an error.

This is a pass-level reproduction using an actual autograd-generated graph, not a full GraphTrainer training launch. Plain `make_fx` does not install GraphTrainer's backward tagging in the tested environment, so the snippet explicitly supplies the backward/recompute metadata expected by the pass. No NPU, CANN, or model assets are required.

Run this script with PyTorch and TorchTitan installed:

```python
"""CPU autograd + make_fx repro of upstream SAR omission; no NPU required."""
import torch
import torch._functorch._activation_offloading.offload_ops
from torch.fx.experimental.proxy_tensor import make_fx
from torch.utils.checkpoint import CheckpointPolicy
from torchtitan.experiments.graph_trainer.selective_activation_remat import selective_activation_remat_pass

lib = torch.library.Library('sar_autograd_review', 'DEF')
lib.define('increment_(Tensor(a!) x) -> ()')
lib.impl('increment_', lambda x: (x.add_(1), None)[1], 'CPU')
lib.impl('increment_', lambda x: None, 'Meta')

class AddOne(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = x.clone()
        torch.ops.sar_autograd_review.increment_(y)
        return y

    @staticmethod
    def backward(ctx, grad):
        return grad

def joint(x):
    y = AddOne.apply(x)
    loss = y.square().sum()
    grad, = torch.autograd.grad(loss, x)
    return loss, grad

x = torch.tensor([2.], requires_grad=True)
gm = make_fx(joint)(x)
# Plain make_fx does not install GraphTrainer's backward tagging in this
# environment. In this deliberately simple graph, everything after loss.sum
# (except output) is autograd-generated backward. Supply the pass metadata
# explicitly, just as in a pass-level regression test.
in_backward = False
for node in gm.graph.nodes:
    if in_backward and node.op != 'output':
        node.meta['autograd_backward'] = True
    if node.target == torch.ops.aten.sum.default:
        in_backward = True
    if node.target == torch.ops.aten.clone.default and not node.meta.get('autograd_backward'):
        node.meta['recompute'] = CheckpointPolicy.MUST_RECOMPUTE

print('before', gm(x))
print('backward-tagged nodes', sum(bool(n.meta.get('autograd_backward')) for n in gm.graph.nodes))
selective_activation_remat_pass(gm)
gm.graph.lint()
print('after SAR', gm(x))
```

Expected behavior: support rematerialization of functional operator wrappers implemented with a private clone followed by a schema-correct in-place kernel. Recomputing the value should preserve the writes required to reconstruct it, so both the forward result and gradients match execution without the rematerialization pass. In this reproduction, the expected loss is 9 and the expected gradient is 6.

We would like GraphTrainer to support this pattern: the wrapper is functional at its public interface, while its implementation uses an in-place kernel on a newly allocated clone. This allows us to use fused operators while retaining activation recomputation. We welcome maintainers' guidance on the appropriate implementation and support scope, including whether to functionalize the graph before SAR or make replay account for the relevant storage dependencies. The proposed direction and eventual support policy are for maintainers to decide; our requested outcome is correct support for this use case.

This was originally observed while using GraphTrainer with a functional partial-RoPE wrapper around an in-place kernel. The CPU example isolates the issue from that hardware/operator implementation. A local experiment that replays the missing write restores the gradient to 6, but a general repair also needs to account for aliases, write ordering, and other write targets. We are reporting the issue first rather than proposing that experiment as a complete fix.

### Versions

- Reproduced on CPU with Python 3.12.10 and `torch==2.14.0+cpu`, on Linux aarch64.
- Rechecked upstream main on September 15, 2026: `f5c97dc0e2d6cca72313004aa853a196ca02ab49` (commit timestamp: 2026-09-15 01:58:39 UTC). Downloaded and directly executed `selective_activation_remat.py` from that exact commit against the CPU reproduction: loss remained 9 and the gradient changed from 6 to 4. The imported `_is_backward_node` and `_get_module_fqn` helpers were also checked against that commit and match. This confirms the defect in the current main SAR pass; it is not a full latest-checkout GraphTrainer training run.
- Upstream source: https://github.com/pytorch/torchtitan/blob/f5c97dc0e2d6cca72313004aa853a196ca02ab49/torchtitan/experiments/graph_trainer/selective_activation_remat.py
- Latest PyTorch/torchao nightly: not independently tested. The reproduction does not invoke torchao operations.
- TOML/config overrides: none; this is a standalone CPU pass-level reproduction.

**Comment by tianyu-l** [comment]
cc @aditvenk @IvanKobzarev
